### PR TITLE
inotify: code cleanup. Closing watcher should now always shut down goroutine

### PR DESCRIPTION
**Pull request by PieterD**
fixes go-fsnotify/fsnotify#3

Also fixed a racey access to w.watches in Close.

I put the tests in notify_test.go, so it doesn't break any other builds. These tests should be checked against the other platforms one by one before pulling them into integration_test.go.

### CODE REVIEW

**Comment by nathany** [comment]
Thanks.

File: inotify.go [hunk 1]
```diff
@@ -101,9 +125,14 @@ func (w *Watcher) Add(name string) error {
 
 // Remove stops watching the the named file or directory (non-recursively).
 func (w *Watcher) Remove(name string) error {
-	name = filepath.Clean(name)
 	w.mu.Lock()
 	defer w.mu.Unlock()
+	return w.remove(name)
+}
+
+// Only call this when w.mu is locked, like from Remove or Close.
```
Comment (nathany) [inline inotify.go]: I'm not in love with having a function that requires the mutex to be done outside of it, but I don't have any better ideas right now.

Also not sure about taking a Mutex around the entirety of remove vs. just some part of it.

File: inotify_test.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
+// Copyright 2010 The Go Authors. All rights reserved.
```
Comment (nathany) [inline inotify_test.go]: © 2015 for new files, old files can rename the same

File: inotify.go (hunk 1 above)
Comment (PieterD) [inline inotify.go]: How about a seperate method to return a slice of watches, or a method to simply remove all watches?

Since the watcher is closed when Closing anyway, no new ones will be added, and we could do more fine-grained locking in remove.

File: inotify_test.go (hunk 2 above)
Comment (PieterD) [inline inotify_test.go]: Got it. Sorry.

File: inotify.go (hunk 1 above)
Comment (PieterD) [inline inotify.go]: I'm not convinced fine-grained locking is an option here. It seems a trap for race conditions at first glance. Maybe it's even a mistake in Add.

**Comment by nathany** [comment]
Have you confirmed that #3 is fixed by this change?

I'll take a closer look and do some QA tomorrow before I merge it in. Thanks again for the help.

**Comment by PieterD** [comment]
I've checked. I did the same as you (checked /proc/x/fd), disabled tests until i found the one that did it. Unsurprisingly, it was a blocking goroutine.

It still happened due to corner case that I hadn't covered yet; in close, when there were watches to remove but they all gave errors, the 'wakeup' wasn't done. Now it is. And thus it is fixed.

**Comment by nathany** [comment]
@PieterD Have you already signed Google's CLA for contributing to Go? (This code base originates from Go exp packages and may one day land in Go).

https://developers.google.com/open-source/cla/individual

**Comment by PieterD** [comment]
Did that (and A+C) years ago. It's fine.

**Comment by PieterD** [comment]
There's still some uncertainty when closing which I'm fixing now.

**Comment by nathany** [comment]
Thanks. I'm on my way to a hack day, fsnotify will be my project for today.

**Comment by PieterD** [comment]
Nice. I'm using epoll to get rid of the last bit of uncertainty.

**Comment by nathany** [comment]
And then it gets interesting. :-)

**Comment by PieterD** [comment]
Yeah, it's a lot of busywork for a small imperfection, but it's got to be done. You could accept the pull as is, and I'll shoot the epoll thing as a seperate pull request. That might ease the review.

**Comment by PieterD** [comment]
Man, it SUCKS that syscall.Read doesn't unblock with you close the file descriptor. Even epoll doesn't respond when you do that.

**Comment by nathany** [comment]
Does IN_NONBLOCK help at all? See #5.

**Comment by nathany** [comment]
I admit I don't understand this stuff. I'm yet to become a Linux guru. :-)

**Comment by nathany** [comment]
A separate PR would be appreciated. Let me get this tested and merged in.

**Comment by PieterD** [comment]
Please. It would be nice to seperate this mess and the next mess, instead of creating one giant ubermess :P

**Comment by nathany** [comment]
übermess!

**Comment by PieterD** [comment]
HOLY CRAP I just figured it out. There's a truly awesome race condition that happens :D

I was wondering how it could be possible that sometimes, test n was making test n+1 fail with "bad file descriptor" or with the wrong number of events.

What happens is that when you've got a goroutine waiting for events, it's possible to close it and create a new one, before other goroutines have a chance to respond. But since the previous descriptor was closed, the new inotify file descriptor will have the same number as the previous one, which will then get closed once the previous test's goroutines wake up and close their stuff.

So, yeah. From now on, Close waits for the readEvents goroutine to finish.

**Comment by nathany** [comment]
Hah. "awesome" race conditions. So awesome that `-race` didn't catch it.

**Comment by PieterD** [comment]
So if, during the execution of Add, some other goroutine closes the watcher and something else opens a file descriptor, it'll get messed up.

This changes a lot. Add and Remove will have to be properly synced with readEvents. Yeah, I don't think think you'll like the next pull request.

**Comment by PieterD** [comment]
I wonder if -race could be adapted to catch this.

File: inotify.go [hunk 3]
```diff
@@ -116,6 +136,21 @@ func (w *Watcher) Remove(name string) error {
 	return nil
 }
 
+// Called from Close(); remove all watches.
+func (w *Watcher) removeAll() int {
+	successful := 0
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	for name, watch := range w.watches {
+		success, _ := syscall.InotifyRmWatch(w.fd, watch.wd)
```
Comment (nathany) [inline inotify.go]: do we need `name = filepath.Clean(name)` here?

It was added to make ./path and path equivalent.

Comment (nathany) [inline inotify.go]: no, sorry, we don't use the name except for delete, so it's fine.

File: inotify.go [hunk 4]
```diff
@@ -149,20 +183,31 @@ func (w *Watcher) readEvents() {
...
-			syscall.Close(w.fd)
-			close(w.Events)
-			close(w.Errors)
+			close(w.done)
 			return
 		}
 
 		if n < 0 {
-			w.Errors <- os.NewSyscallError("read", errno)
+			select {
```
Comment (nathany) [inline inotify.go]: these two selects make me curious.

we are sending an error or checking if the done channel is closed, whichever happens first?

File: inotify.go (hunk 3 above)
Comment (PieterD) [inline inotify.go]: We don't need removeAll at all, actually. closing the fd will get rid of everything.

Comment (PieterD) [inline inotify.go]: In fact, we don't need watches at all.

File: inotify.go (hunk 4 above)
Comment (nathany) [inline inotify.go]: so if the consumer of this library isn't ready to receive an Error or Event for some reason, we don't continue blocking if Close() is called.,

File: inotify_test.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+func TestInotifyCloseSlightlyLaterWithWatch(t *testing.T) {
+	testDir := tempMkdir(t)
+	defer os.RemoveAll(testDir)
+
+	w, err := NewWatcher()
+	if err != nil {
+		t.Fatalf("Failed to create watcher")
+	}
+	w.Add("testDir")
```
Comment (nathany) [inline inotify_test.go]: :bug: testDir without quotes?

File: inotify_test.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+func TestInotifyCloseSlightlyLater(t *testing.T) {
+	w, err := NewWatcher()
+	if err != nil {
+		t.Fatalf("Failed to create watcher")
+	}
+
+	// Wait until readEvents has reached syscall.Read, and Close.
+	<-time.After(50 * time.Millisecond)
```
Comment (nathany) [inline inotify_test.go]: Is there any advantage/difference of time.After vs. time.Sleep?

File: inotify.go (hunk 4 above)
Comment (PieterD) [inline inotify.go]: Yeah. This solves the problem of when someone Closes the watcher and moves on, but there is still an error waiting to be sent. Without those selects, it would block forever. Which is, in fact, what was happening in a lot of cases.

Comment (PieterD) [inline inotify.go]: Er, anyway it still blocks on sending the error or event, unless we closed it. Only then will w.done return a value.

File: inotify_test.go (hunk 5 above)
Comment (PieterD) [inline inotify_test.go]: Heh. OOPS

Comment (nathany) [inline inotify_test.go]: I can fix it, don't worry about it.

Comment (PieterD) [inline inotify_test.go]: Ok, thanks.

**Comment by nathany** [comment]
For the changelog, #5 isn't actually solved yet, correct? (coming next)

**Comment by PieterD** [comment]
I was under the impression that https://github.com/go-fsnotify/fsnotify/issues/5 was about goroutines blocking forever.

Perhaps I'm mistaken.

File: inotify.go (hunk 3 above)
Comment (nathany) [inline inotify.go]: Don't we still need to remove things from our internal map? Otherwise a later Add will think it's already being watched.

File: inotify_test.go (hunk 6 above)
Comment (PieterD) [inline inotify_test.go]: Not really.

**Comment by nathany** [comment]
Squashed and merged. 

#5. Doesn't readEvent still block though? That little hack to wake it up on Close might work, but maybe not?

File: inotify.go (hunk 3 above)
Comment (PieterD) [inline inotify.go]: We only use removeAll when we've already closed.

Ah, and we do need watches for the inotify handles. Sorry.

**Comment by PieterD** [comment]
readEvents still has a blocking read on the file descriptor, but that's fine. So does kqueue and the Windows thing.

It should get cleaned up pretty much always, but not ABSOLUTELY always. That's the next one. So yeah, I suppose #5 might not be fixed until I push the epoll stuff.

**Comment by nathany** [comment]
kqueue has a timeout right now so that it doesn't block indefinitely, but it does use a little bit of CPU to go around the loop every few 100ms.
